exp/zapslog/slog.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2023 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:build zapslog
+
+package zapslog
+
+import (
+	"context"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"golang.org/x/exp/slog"
+)
+
+type handler struct {
+	logger *zap.Logger
+}
+
+var _ slog.Handler = (*handler)(nil)
+
+// group holds all the Attrs saved in a slog.GroupValue.
+type group struct {
+	attrs []slog.Attr
+}
+
+func (g *group) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	for _, attr := range g.attrs {
+		convertAttrToField(attr).AddTo(enc)
+	}
+	return nil
+}
+
+func convertAttrToField(attr slog.Attr) zapcore.Field {
+	switch attr.Value.Kind() {
+	case slog.KindBool:
+		return zap.Bool(attr.Key, attr.Value.Bool())
+	case slog.KindDuration:
+		return zap.Duration(attr.Key, attr.Value.Duration())
+	case slog.KindFloat64:
+		return zap.Float64(attr.Key, attr.Value.Float64())
+	case slog.KindInt64:
+		return zap.Int64(attr.Key, attr.Value.Int64())
+	case slog.KindString:
+		return zap.String(attr.Key, attr.Value.String())
+	case slog.KindTime:
+		return zap.Time(attr.Key, attr.Value.Time())
+	case slog.KindUint64:
+		return zap.Uint64(attr.Key, attr.Value.Uint64())
+	case slog.KindGroup:
+		return zap.Object(attr.Key, &group{attrs: attr.Value.Group()})
+	case slog.KindLogValuer:
+		return convertAttrToField(slog.Attr{
+			Key:   attr.Key,
+			Value: attr.Value.Resolve(),
+		})
```

Comment:
This will resolve the LogValuer eagerly, which will hurt anyone using LogValuer to defer expensive operations.

Ideally, we want to use a wrapper struct that calls LogValue() at serialization time,
but that may not be 100% straightforward because a LogValuer doesn't know if it's a group or a primitive type until calling LogValue(),
whereas Zap wants to know that in advance.
This may require a lazy field-equivalent in Zap.
We can defer designing that to a follow-up. 
CC @mway @prashantv @sywhang

Reply:
+1 on a adding a lazy-field equivalent in zap separately.